src/components/deck/index.js

```diff
@@ -1,36 +1,52 @@
-import React, { useCallback, useMemo, useState } from 'react';
+import React, { useCallback, useRef } from 'react';
 import PropTypes from 'prop-types';
-import { parse as parseQS } from 'query-string';
+import { parse as parseQS, stringify as stringifyQS } from 'query-string';
 import DefaultDeck from './default-deck';
 import PresenterMode from '../presenter-mode';
+import PrintMode from '../../print-mode';
 import useMousetrap from '../../hooks/use-mousetrap';
 import { KEYBOARD_SHORTCUTS, SPECTACLE_MODES } from '../../utils/constants';
+import { modeKeyForSearchParam, modeSearchParamForKey } from './modes';
 
 export default function SpectacleDeck(props) {
-  const { search: queryString } = location;
-  const { presenterMode, overviewMode } = parseQS(queryString, {
-    parseBooleans: true
-  });
-
-  const defaultMode = useMemo(() => {
-    if (presenterMode) {
-      return SPECTACLE_MODES.PRESENTER_MODE;
-    } else if (overviewMode) {
-      return SPECTACLE_MODES.OVERVIEW_MODE;
-    }
-    return SPECTACLE_MODES.DEFAULT_MODE;
-  }, [overviewMode, presenterMode]);
-
-  const [mode, setMode] = useState(defaultMode);
+  const mode = useRef(
+    modeKeyForSearchParam(
+      parseQS(location.search, {
+        parseBooleans: true
+      })
+    )
+  );
 
   const toggleMode = useCallback(
-    (e, newMode) => {
+    (e, newMode, senderSlideIndex) => {
       e?.preventDefault();
-      if (mode === newMode) {
-        setMode(SPECTACLE_MODES.DEFAULT_MODE);
+
+      let stepIndex = 0;
+      let slideIndex = senderSlideIndex;
+      const searchParams = parseQS(location.search, {
+        parseBooleans: true
+      });
+
+      if (!slideIndex) {
+        slideIndex = searchParams.slideIndex;
+        stepIndex = searchParams.stepIndex;
+      }
+
+      if (mode.current === newMode) {
+        location.search = stringifyQS({
+          slideIndex,
+          stepIndex
+        });
         return;
       }
-      setMode(newMode);
+
+      mode.current = newMode;
+
+      location.search = stringifyQS({
+        slideIndex,
+        stepIndex,
+        ...modeSearchParamForKey(newMode)
```

Comment:
For every mode transition we will update the `location` object so it will be also tracked in history. This also preserves existing params as well.